# P4 Backward propogation
Voor deze opdracht moeten we een neural network maken met backward propogation. Hieronder Importeer ik alle bestanden waar dit in gedaan wordt en test ik vervolgens de AND en XOR gate en de Half Adder

In [1]:
from P4.code.neuron import Neuron
from P4.code.neuronlayer import NeuronLayer
from P4.code.neuronnetwerk import NeuronNetwork

We gaan de gates testen door middel van de voorbeelden in het werkboek van les 6

### AND Gate

De AND Gate zag er als volgt uit:
<img src="./images/and.png" alt="image" width="400"/>

De values waren als volgt:

<img src="./images/and_values.png" alt="image" width="300"/>

Met deze info zagen de iteraties er als volgt uit (we doen maar 1 epoch dus we hebben 4 iteraties):

<img src="./images/and_it1.png" alt="image" width="500"/>
<img src="./images/and_it2.png" alt="image" width="500"/>
<img src="./images/and_it3.png" alt="image" width="500"/>
<img src="./images/and_it4.png" alt="image" width="500"/>


Hieronder maken we de neuron aan en tonen we wat wij krijgen uit onze AND gate

In [2]:
n1 = Neuron([-0.5, 0.5], 1.5)
l1 = NeuronLayer([n1])
network = NeuronNetwork([l1], 1)

network.train([[0,0],[1,0],[0,1],[1,1]],[0,0,0,1], 1)


Epoch 1: returns [0.818] with an input of [0, 0] and a loss of 0.251
VALUES AFTER UPDATE: weights: [-0.5, 0.5], error: 0.122, bias: 1.378

Epoch 1: returns [0.706] with an input of [1, 0] and a loss of 0.241
VALUES AFTER UPDATE: weights: [-0.647, 0.5], error: 0.147, bias: 1.232

Epoch 1: returns [0.85] with an input of [0, 1] and a loss of 0.225
VALUES AFTER UPDATE: weights: [-0.647, 0.391], error: 0.109, bias: 1.123

Epoch 1: returns [0.704] with an input of [1, 1] and a loss of 0.214
VALUES AFTER UPDATE: weights: [-0.585, 0.453], error: -0.062, bias: 1.185


Zoals je ziet klopt alles. Het enige wat niet hetzelfde is is de loss, maar dit is een fout in het werkboek (als het goed is). Want aan het begin wordt getoond dat de loss bij [0,0] 0.251 is, maar bij de eerste iteratie staat er opeens dat het 0.241 is.

<img src="./images/and_correct.png" alt="image" width="500"/>

Ik heb zelf met de hand ook gerekend en kwam op hetzelfde antwoord uit als hier. Mijn code kwam ook hierop. Dus ik denk dat alles eentje omlaag moet. [0,0] heeft een loss van 0.251, [1,0] heeft een loss van 0.241 ipv 0.225, [0,1] heeft 0.225 ipv 0.214. enz.

### XOR Gate

De XOR Gate zag er als volgt uit:
<img src="./images/xor.png" alt="image" width="400"/>

De values waren als volgt:

<img src="./images/xor_values.png" alt="image" width="300"/>

Met deze info zagen de iteraties er als volgt uit (we doen maar 1 epoch, maar in het werkboek worden alleen de eerste 2 iteraties getoond):

<img src="./images/xor_it1.png" alt="image" width="500"/>
<img src="./images/xor_it2.png" alt="image" width="500"/>



Hieronder de xor gate. We gaan alle inputs en targets meegeven, maar om te vergelijken moet je naar de eerste 2 kijken.

In [3]:
f = Neuron([0.2, -0.4], 0)
g = Neuron([0.7, 0.1], 0)

layer_1 = NeuronLayer([f, g])

o = Neuron([0.6, 0.9], 0)

layer_2 = NeuronLayer([o])

network = NeuronNetwork([layer_1, layer_2], 1)

network.train([[1,1], [0,1], [1,0], [0,0]], [0, 1, 1, 0], 1)


Epoch 1: returns [0.709] with an input of [1, 1] and a loss of 0.144
VALUES AFTER UPDATE: weights: [0.186, -0.414], error: 0.014, bias: -0.014
VALUES AFTER UPDATE: weights: [0.68, 0.08], error: 0.02, bias: -0.02
VALUES AFTER UPDATE: weights: [0.534, 0.799], error: 0.146, bias: -0.146

Epoch 1: returns [0.617] with an input of [0, 1] and a loss of 0.135
VALUES AFTER UPDATE: weights: [0.186, -0.428], error: 0.014, bias: -0.028
VALUES AFTER UPDATE: weights: [0.68, 0.056], error: 0.023, bias: -0.044
VALUES AFTER UPDATE: weights: [0.57, 0.846], error: -0.091, bias: -0.056

Epoch 1: returns [0.691] with an input of [1, 0] and a loss of 0.139
VALUES AFTER UPDATE: weights: [0.171, -0.428], error: 0.014, bias: -0.043
VALUES AFTER UPDATE: weights: [0.658, 0.056], error: 0.022, bias: -0.065
VALUES AFTER UPDATE: weights: [0.605, 0.889], error: -0.066, bias: 0.01

Epoch 1: returns [0.676] with an input of [0, 0] and a loss of 0.143
VALUES AFTER UPDATE: weights: [0.171, -0.428], error: 0.014, bias:

Zoals je ziet zijn bepaalde waardes net ernaast, ik weet niet 100% waardoor dit komt want de berekeningen kloppen wel. Misschien is het door de round dat ie het net anders afrond in de print. Desondanks kloppen de outputs wel.

### Half Adder

De Half Adder ziet er als volgt uit:
<img src="./images/ha.png" alt="image" width="400"/>

De values waren als volgt:

<img src="./images/ha-values.png" alt="image" width="300"/>

Er wordt maar 1 iteratie getoond met een loss. Deze gaan we ook alleen testen omdat de half adder veel print teruggeeft.
<img src="./images/hu_ff.png" alt="image" width="400"/>
<img src="./images/ha_update.png" alt="image" width="400"/>

In [4]:
f = Neuron([0, 0.1], 0)
g = Neuron([0.2, 0.3], 0)
h = Neuron([0.4, 0.5], 0)
# carry
c = Neuron([0.9, 1.0, 1.1], 0)
# sum
s = Neuron([0.6, 0.7, 0.8], 0)

layer_1 = NeuronLayer([f, g, h])
layer_2 = NeuronLayer([s, c])

network = NeuronNetwork([layer_1, layer_2], 1)

network.train([[1, 1]], [[0, 1]], 1)


Epoch 1: returns [0.789, 0.867] with an input of [1, 1] and a loss of 0.334
VALUES AFTER UPDATE: weights: [0.0, 0.1], error: -0.0, bias: 0.0
VALUES AFTER UPDATE: weights: [0.195, 0.295], error: 0.005, bias: -0.005
VALUES AFTER UPDATE: weights: [0.391, 0.491], error: 0.009, bias: -0.009
VALUES AFTER UPDATE: weights: [0.531, 0.618, 0.707], error: 0.131, bias: -0.131
VALUES AFTER UPDATE: weights: [0.908, 1.01, 1.111], error: -0.015, bias: 0.015


De dit is interessant. De output weights en biases en errors zijn goed, maar de loss is anders. De geupdate weights van g en h zijn ook goed, maar die van f zijn weer anders. En bij zowel f,g en h zijn de biases fout, maar ondanks dat die fout zijn is de output dus wel helemaal correct. Misschien is de bias anders door de learning rate. Deze is namelijk niet aangegeven bij het begin dus ging ik ervan uit dat het 1 was aangezien het ook zo was bij de XOR en AND gate.

Dataset:

Aangezien ik redelijk wat moet inleveren, waaronder nog P5 heb ik besloten de dataset tests niet uit te voeren.